base/abstractarray.jl

```diff
@@ -3340,6 +3344,9 @@ end
 Like [`map`](@ref), but stores the result in `destination` rather than a new
 collection. `destination` must be at least as large as the smallest collection.
 
+The order in which the elements of individual `collection[i]` arguments are iterated is undefined.
+If `destination` shares memory with any of the `collection[i]` arguments, the result is undefined.
```

Comment:
This is undefined?
```julia
x = rand(100)
map!(sqrt, x, x)
```

Reply:
I mean it could be right? consider something pathological
```
abstract type LinkedArrayEntry end
struct NullEntry <: LinkedArrayEntry
    val
    NullEntry() = new(0)
end

mutable struct LeftLinkedArrayEntry <: LinkedArrayEntry
    left::LinkedArrayEntry
    val
end

x1 = NullEntry()
x2 = LeftLinkedArrayEntry(x1, 2)
x3 = LeftLinkedArrayEntry(x2, 3)
x4 = LeftLinkedArrayEntry(x3, 4)

x = LinkedArrayEntry[x1, x2, x3, x4]

lsum(::NullEntry) = NullEntry()
lsum(l::LeftLinkedArrayEntry) = LeftLinkedArrayEntry(l.left, l.val + l.left.val)

### always gives new vector with sums of adjacent values
y = Vector{LinkedArrayEntry}(undef, 4)
for i ∈ 1:4
    y[i] = lsum(x[i])
end

### undefined
map!(lsum, x, x)
```

I think the warning is basically saying "hey, `map!` doesn't copy for you or do any safety checks whatsoever when inputs alias"

however, maybe "undefined" is the wrong word, since it's not UB in the C sense?

Reply:
For the moment, for simplicity, let's assume there is only one input collection. In that case, the following sentence would technically be sufficient:

```
The order in which the collection is iterated is undefined.
```

Then it follows that `map!(sqrt, x, x)` produces the correct result, since the iteration order doesn't matter for that call. It also follows that weird aliasing situations might produce incorrect results, but it's not quite obvious from that sentence alone.

...On a previous iteration of this PR, we had the following:

```
The order in which the `collection`s are iterated is undefined. If `destination` shares memory
with any of the `collection`s, the result might be incorrect.
```

That's close, because of the word "might", but still not ideal, because we want the correctness of `map!(sqrt, x, x)` to follow from the docstring, which it does if the only sentence is "The order in which the collection is iterated is undefined."

Reply:
but in theory, couldn't there be pathological situations where the order of iterations matter iff the inputs alias?

Reply:
that's what the current version of the PR assumes